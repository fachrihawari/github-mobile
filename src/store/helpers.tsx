import { Action as ReduxAction } from "redux";

export interface Action<T extends string, P> extends ReduxAction<T> {
  payload: P;
}

export function createAction<T extends string>(type: T): Action<T, void>;
export function createAction<T extends string, P>(
  type: T,
  payload: P
): Action<T, P>;
export function createAction<T extends string, P>(type: T, payload?: P) {
  return typeof payload === "undefined" ? { type } : { type, payload };
}
