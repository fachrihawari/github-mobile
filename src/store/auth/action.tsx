import { FETCH_USER_REQUEST, LOGOUT } from "./constant";
import { createAction } from "../helpers";

export const fetchUserRequest = (username: string, password: string, OTP: string='') =>
  createAction(FETCH_USER_REQUEST, { username, password, OTP });

export const logout = () => createAction(LOGOUT);
