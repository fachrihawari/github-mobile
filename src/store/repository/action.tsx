import { FETCH_COMMIT_REQUEST } from "./constant";
import { createAction } from "../helpers";

export const fetchCommitRequest = (
  repository: string,
  page?: number,
  perPage?: number
) => createAction(FETCH_COMMIT_REQUEST, { repository, page, perPage });
