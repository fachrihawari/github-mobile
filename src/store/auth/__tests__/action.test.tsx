import { parseError } from '../../helpers';
import * as action from '../action';
import * as constant from '../constant';

describe('auth actions', () => {
  it('should create an action to fetch user profile', () => {
    const credentials = {
      username: 'Username',
      password: 'Password',
      OTP: ''
    }
    const fetchUserRequestAction = action.fetchUserRequest(credentials.username, credentials.password)
    const expectedAction = {
      type: constant.FETCH_USER_REQUEST,
      payload: credentials
    }
    expect(fetchUserRequestAction).toStrictEqual(expectedAction)
    expect(fetchUserRequestAction).toMatchSnapshot()
  })

  it('should create an action to fetch user profile if the user enabled 2FA', () => {
    const credentials = {
      username: 'Username',
      password: 'Password',
      OTP: '123456'
    }
    const fetchUserRequestAction = action.fetchUserRequest(credentials.username, credentials.password, credentials.OTP)
    const expectedAction = {
      type: constant.FETCH_USER_REQUEST,
      payload: credentials
    }
    expect(fetchUserRequestAction).toStrictEqual(expectedAction)
    expect(fetchUserRequestAction).toMatchSnapshot()
  })

  it('should create an action to set loader when fetching user', () => {
    const expectedAction = {
      type: constant.FETCH_USER_BEGIN
    }
    expect(action.fetchUserBegin()).toStrictEqual(expectedAction)
    expect(action.fetchUserBegin()).toMatchSnapshot()
  })

  it('should create an action to set profile user when success fetching user', () => {
    const profile = {
      id: 1,
      name: 'User'
    }
    const expectedAction = {
      type: constant.FETCH_USER_SUCCESS,
      payload: {
        profile
      }
    }
    expect(action.fetchUserSuccess(profile)).toStrictEqual(expectedAction)
    expect(action.fetchUserSuccess(profile)).toMatchSnapshot()
  })


  it('should create an action to set error message when failure fetching user because bad credentials', () => {
    const error = {
      response: {
        data: {
          message: 'Bad credentials'
        }
      }
    }
    const expectedAction = {
      type: constant.FETCH_USER_FAILURE,
      payload: {
        error: parseError(error),
        needOTP: false
      }
    }
    expect(action.fetchUserFailure(parseError(error), false)).toStrictEqual(expectedAction)
    expect(action.fetchUserFailure(parseError(error), false)).toMatchSnapshot()
  })

  it('should create an action to set error message when failure fetching user because network problem', () => {
    const error = 'Lost connection'
    const expectedAction = {
      type: constant.FETCH_USER_FAILURE,
      payload: {
        error: parseError(error),
        needOTP: false
      }
    }
    expect(action.fetchUserFailure(parseError(error), false)).toStrictEqual(expectedAction)
    expect(action.fetchUserFailure(parseError(error), false)).toMatchSnapshot()
  })

  it('should create an action to logout', () => {
    const logoutAction = action.logout()
    const expectedAction = {
      type: constant.LOGOUT
    }
    expect(logoutAction).toStrictEqual(expectedAction)
    expect(logoutAction).toMatchSnapshot()
  })
})