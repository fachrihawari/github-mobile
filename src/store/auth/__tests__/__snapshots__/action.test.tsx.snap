// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth actions should create an action to fetch user profile 1`] = `
Object {
  "payload": Object {
    "OTP": "",
    "password": "Password",
    "username": "Username",
  },
  "type": "auth/FETCH_USER_REQUEST",
}
`;

exports[`auth actions should create an action to fetch user profile if the user enabled 2FA 1`] = `
Object {
  "payload": Object {
    "OTP": "123456",
    "password": "Password",
    "username": "Username",
  },
  "type": "auth/FETCH_USER_REQUEST",
}
`;

exports[`auth actions should create an action to logout 1`] = `
Object {
  "type": "auth/LOGOUT",
}
`;

exports[`auth actions should create an action to set error message when failure fetching user because bad credentials 1`] = `
Object {
  "payload": Object {
    "error": "Bad credentials",
    "needOTP": false,
  },
  "type": "auth/FETCH_USER_FAILURE",
}
`;

exports[`auth actions should create an action to set error message when failure fetching user because network problem 1`] = `
Object {
  "payload": Object {
    "error": "You're offline!",
    "needOTP": false,
  },
  "type": "auth/FETCH_USER_FAILURE",
}
`;

exports[`auth actions should create an action to set loader when fetching user 1`] = `
Object {
  "type": "auth/FETCH_USER_BEGIN",
}
`;

exports[`auth actions should create an action to set profile user when success fetching user 1`] = `
Object {
  "payload": Object {
    "profile": Object {
      "id": 1,
      "name": "User",
    },
  },
  "type": "auth/FETCH_USER_SUCCESS",
}
`;
